_posts/2019-02-05-configurable-builds-part-1.md

```diff
@@ -0,0 +1,280 @@
+
+```sh
+$ bazel build //my:cc_binary --cpu=x86
+$ bazel build //my:cc_binary --cpu=arm
+$ bazel build //my:android_binary --android_sdk=@androidsdk//:sdk-28
+```
+
+This post reviews Bazel's *configurability work*, which is an ongoing effort to
+make these tasks simple, flexible, and powerful. This is a deeper dive into
+themes covered in Bazel's [configurability
+roadmap](https://bazel.build/roadmaps/configuration.html).
+
+## History
+
+Bazel's a
+[powerful](https://docs.bazel.build/versions/master/bazel-vision.html) build
+tool that's especially suited for large,
+[multi-language](/2018/12/05/multilanguage-build-system.html) builds. But it
+was born out of Google, which historically wrote code for homogeneous fleets of
+Linux servers with little need for customization. What customization was needed
+was achieved with ad hoc flags and logic built
+[straight](https://source.bazel.build/bazel/+/144912e7b7a86b45e07f79e76f6fed20890acb36:src/main/java/com/google/devtools/build/lib/rules/cpp/CppOptions.java;l=258)
+into the build tool.
+
+This is no longer true inside or outside of Google. Modern software targets
+phones, cloud, servers, desktops, [smart
+devices](https://en.wikipedia.org/wiki/Smart_device) and more, and must offer
+increasingly flexible support for any intended combination of these platforms.
+
+This means
+
+    $ bazel build //my:binary
+
+isn't enough to describe what you want to do. What platform(s) are you
+targeting? What features do you want to include? What if you want to build a
+complex app with generated sources, client and server modules, native
+extensions, and test data?
+
+Bazel's historical approach of ad hoc flags:
+
+    $ bazel build //my:binary --cpu=arm --crosstool_top=//my:custom_toolchain --define MYFEATURE=1
+
+is lacking in many ways:
+
+1. `--cpu` only accepts values explicitly supported by C++ rules. Even if you're
+   not building C++.
+1. `--crosstool_top`, which specifies exactly what command line should compile
+   your code, is even more tied to C++. Other languages may be organized
+   completely differently.
+1. `--define` is completely unstructured and unreadable in Starlark.
+1. `$ bazel build //my:all` sets the same flags for all targets, even if
+   different targets need different flags.
+1. `//my:binary`'s deps must use the same settings as their parent.
+1. It might be hard to remember which flags are needed by which targets,
+   especially across users and dev/test/prod environments. This risks
+   inconsistent and possibly incorrect builds.
+
+A different approach might be to set flags directly in targets that need them,
+let rule writers design the flags that affect their rules, and provide standard
+extensible APIs for cross-language concepts like *platform* and *cpu*. This
+lets people configure their projects on their own terms, using language
+consistent with how their projects are organized.
+
+Bazel's configurability work is redesigning Bazel around these ideas. It's
```

Comment:
Editorial suggestion: I think you could start with this as the hook of the article. Rather than building up to this point slowly over the course of the article, you could start with "One of Bazel's ultimate goals is to make `bazel build //:all` just work for any configuration, platform, toolchain, or language you might be using. I want to share how we're planning to get there with our recent configurability redesign—along with the problems we've faced getting to this point."

Leading with the key goal is really helpful for a reader who wants to know if they should read the rest of the article and can give a nice framework for the rest of the article.

What do you think?

Reply:
Great suggestion - I think I've figured out a way to make this work.  This is also important for the blog index / post preview, which only includes the first paragraph. That really wasn't punchy with the original lead.